jupyter_core/paths.py

```diff
@@ -9,21 +9,78 @@
 
 
 import os
+import ast
 import sys
 import stat
 import errno
+import time
+import pathlib
 import tempfile
 import warnings
+import functools
+import traceback
+import importlib
 
 from contextlib import contextmanager
 
+import entrypoints
+
 pjoin = os.path.join
 
 # UF_HIDDEN is a stat flag not defined in the stat module.
 # It is used by BSD to indicate hidden files.
 UF_HIDDEN = getattr(stat, 'UF_HIDDEN', 32768)
 
 
+# The entry_point MUST be the name of a directory, consisting of only
+# letters, numbers, and underscores, adjacent to __init__.py
+JUPYTER_DATA_PATH_ENTRY_POINT = "jupyter_data_path"
+JUPYTER_CONFIG_PATH_ENTRY_POINT = "jupyter_config_path"
+
+# TODO: decide whether to keep, or grow a logging endpoint
+JUPYTER_ENTRY_POINT_TIMINGS = os.environ.get("JUPYTER_ENTRY_POINT_TIMINGS")
+
+
+def _get_path_from_one_entry_point(ep):
+    """ use the entrypoint attribute name to discover the path without loading
+    """
+    spec = importlib.util.find_spec(ep.module_name)
+    module = importlib.util.module_from_spec(spec)
+    origin = pathlib.Path(module.__file__).parent.resolve()
```

Comment:
`module.__file__` is `None` if there's no top-level `__init__.py` file in the module